File: backend/src/main/java/moe/mewore/rabbit/backend/simulation/data/FrameCompiler.java
package moe.mewore.rabbit.backend.simulation.data;

import java.util.Arrays;
import java.util.stream.StreamSupport;

import lombok.Getter;

public class FrameCompiler {

    @Getter
    private int size = 0;

    public FrameSection reserve(final FrameDataType... elementTypes) {
        return reserve(Arrays.stream(elementTypes).mapToInt(FrameDataType::getByteSize).sum());
    }

    public FrameSection reserve(final Iterable<FrameDataType> elementTypes) {
        return reserve(
            StreamSupport.stream(elementTypes.spliterator(), true).mapToInt(FrameDataType::getByteSize).sum());
    }

    private FrameSection reserve(final int reservationSize) {
        final FrameSection result = new FrameSection(size, reservationSize);
        size += reservationSize;
        return result;
    }

    public FrameSection[] reserveMultiple(final int count, final FrameDataType... elementTypes) {
        final FrameSection[] result = new FrameSection[count];
        for (int i = 0; i < count; i++) {
            result[i] = reserve(elementTypes);
        }
        return result;
    }
}

File: frontend/src/game/simulation/data/frame-compiler.ts
import { FrameDataType } from './frame-data-type';
import { FrameSection } from './frame-section';

export class FrameCompiler {
    private size = 0;

    reserve(...elementTypes: FrameDataType[]): FrameSection {
        const reservationSize = elementTypes.map((type) => type.byteSize).reduce((prev, cur) => prev + cur, 0);
        const result = new FrameSection(this.size, reservationSize);
        this.size += reservationSize;
        return result;
    }

    public reserveMultiple(count: number, ...elementTypes: FrameDataType[]): FrameSection[] {
        const result: FrameSection[] = [];
        for (let i = 0; i < count; i++) {
            result.push(this.reserve(...elementTypes));
        }
        return result;
    }
}
